app/routes/dashboard/profile/edit.tsx

```diff
@@ -0,0 +1,243 @@
+import type { User } from '@prisma/client'
+import type { ActionFunction, LoaderFunction } from 'remix'
+import {
+  redirect,
+  useActionData,
+  useTransition,
+  Form,
+  json,
+  useLoaderData,
+} from 'remix'
+import { auth } from '~/services/auth.server'
+import { Button, Field } from '~/components/form-elements'
+import { validatePhoneNumber, validateRequired } from '~/utils/validators'
+import { db } from '~/utils/db.server'
+import { getUser } from '~/models/user'
+import { logout } from '~/services/session.server'
+
+export const loader: LoaderFunction = async ({ request }) => {
+  // If the user is here, it's already authenticated, if not redirect them to
+  // the login page.
+  const { id } = await auth.isAuthenticated(request, {
+    failureRedirect: '/login',
+  })
+
+  // Get the user data from the database.
+  const user = await getUser(id)
+  if (!user) {
+    return logout(request)
+  }
+  return json({ user })
+}
```

Comment:
For loading the user data, you can literally copy this code block and paste it into your `profile/index.tsx` file.

Then in the component, you can simply use this hook to get the user data:
```
const { user } = useLoaderData<{ user: User }>()
```
That way, you can access the current user's data from the `user` object.

Reply:
Oh, I see, you use remix-auth here right.

Reply:
Yes, I use `remix-auth` to get the user ID from the session cookie, then I call `getUser(id)` to get the latest data from the database.

Reply:
what about the e2e test results here mas @zainfathoni, do you have any suggestions?
```6 failed
    [Pixel 4] › e2e/profile.spec.ts:8:1 › Dashboard page has Profile nav link in side menu =========
    [Pixel 4] › e2e/profile.spec.ts:18:1 › Profile page has rendered in dashboard page after click Profile in side menu 
    [Pixel 4] › e2e/profile.spec.ts:31:1 › Redirect to /dashboard/profile/edit after click Ubah button 
    [iPhone 11] › e2e/profile.spec.ts:8:1 › Dashboard page has Profile nav link in side menu =======
    [iPhone 11] › e2e/profile.spec.ts:18:1 › Profile page has rendered in dashboard page after click Profile in side menu 
    [iPhone 11] › e2e/profile.spec.ts:31:1 › Redirect to /dashboard/profile/edit after click Ubah button ```

Reply:
Oh, the errors only happen in the mobile view. It means that they are related to mobile view styling.
In this case, it fails to find the link because it hides the links until the user clicks the `open sidebar` button.

You can add [this additional step](https://github.com/zainfathoni/kelas.rumahberbagi.com/blob/0a0a37e495e64421c47b8be67f7e7cf1cd3e836b/e2e/logout.spec.ts#L12-L17) into your test to cater for the mobile view interaction.

Reply:
after adding the mobile view test in the test case it seems the code still failing, not quite sure where it's failed though. 
```
15 failed
    [Pixel 4] › e2e/profile.spec.ts:8:1 › Dashboard page has Profile nav link in side menu =========
    [Pixel 4] › e2e/profile.spec.ts:37:1 › Profile page has rendered in dashboard page after click Profile in side menu 
    [Pixel 4] › e2e/profile.spec.ts:73:1 › Redirect to /dashboard/profile/edit after click Ubah button 
    [iPhone 11] › e2e/profile.spec.ts:8:1 › Dashboard page has Profile nav link in side menu =======
    [iPhone 11] › e2e/profile.spec.ts:37:1 › Profile page has rendered in dashboard page after click Profile in side menu 
    [iPhone 11] › e2e/profile.spec.ts:73:1 › Redirect to /dashboard/profile/edit after click Ubah button 
    [firefox] › e2e/profile.spec.ts:8:1 › Dashboard page has Profile nav link in side menu =========
    [firefox] › e2e/profile.spec.ts:37:1 › Profile page has rendered in dashboard page after click Profile in side menu 
    [firefox] › e2e/profile.spec.ts:73:1 › Redirect to /dashboard/profile/edit after click Ubah button 
    [webkit] › e2e/profile.spec.ts:8:1 › Dashboard page has Profile nav link in side menu ==========
    [webkit] › e2e/profile.spec.ts:37:1 › Profile page has rendered in dashboard page after click Profile in side menu 
    [webkit] › e2e/profile.spec.ts:73:1 › Redirect to /dashboard/profile/edit after click Ubah button 
    [noscript] › e2e/profile.spec.ts:8:1 › Dashboard page has Profile nav link in side menu ========
    [noscript] › e2e/profile.spec.ts:37:1 › Profile page has rendered in dashboard page after click Profile in side menu 
    [noscript] › e2e/profile.spec.ts:73:1 › Redirect to /dashboard/profile/edit after click Ubah button 
  5 skipped
  35 passed (4m)
```

Reply:
Have you tried playing around with [Playwright's codegen](https://playwright.dev/docs/codegen#generate-tests) feature to ensure that you're writing the correct code expected by Playwright for the actions you want to make?

Reply:
I will try to make some time to debug your test later.

Reply:
> Have you tried playing around with [Playwright's codegen](https://playwright.dev/docs/codegen#generate-tests) feature to ensure that you're writing the correct code expected by Playwright for the actions you want to make?

No, not yet. I'll try it later.

Reply:
I know what your mistake is. You should only **insert steps specifically for mobile** instead of **adding a whole new test set for the mobile-only cases**. I'll give you an example in a bit.

Reply:
Another mistake in this implementation is, somehow it didn't render the sidebar accessibly so that it can get the button through the `getByRole` selector.

You can experiment with the [Testing Playground](https://testing-playground.com)'s browser extensions as I did here:

https://user-images.githubusercontent.com/6315466/147620279-54b19e66-5aaf-4ec1-8420-4593c3d9f3c9.mov

Reply:
This demo shows the expected behaviour of that navigation item.

https://user-images.githubusercontent.com/6315466/147620690-02a8b59a-8e5c-4817-8ab6-e13943ecdb37.mov

Anyway, this issue is out of the #71 issue's scope. We should do it within the #35 issue instead.
Therefore, let's wait for #35. Instead, you can focus on building and testing the profile route and skip all the testing related to the page navigations.

Reply:
I just closed #35 in #81. Please update from the latest `main` branch, it should be more obvious for you now.